// pix-qr-code-detail/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PixQrCodeParseError = void 0;
exports.parseTLV = parseTLV;
exports.extractPixUrl = extractPixUrl;
exports.decodeJwt = decodeJwt;
exports.getDecodedPixJwt = getDecodedPixJwt;
const ROOT_NAMES = {
    '00': 'Payload Format Indicator',
    '01': 'Point of Initiation Method',
    '52': 'Merchant Category Code',
    '53': 'Transaction Currency',
    '54': 'Transaction Amount',
    '58': 'Country Code',
    '59': 'Merchant Name',
    '60': 'Merchant City',
    '61': 'Postal Code',
    '62': 'Additional Data Field Template',
    '63': 'CRC-16',
    '64': 'Merchant Information - Language Template'
};
const MAI_NAMES = {
    '00': 'Globally Unique Identifier',
    '01': 'Chave Pix',
    '02': 'Informacoes Adicionais',
    '25': 'URL do Payload'
};
const ADD_DATA_NAMES = {
    '05': 'Reference Label'
};
const PIX_GUI = 'BR.GOV.BCB.PIX';
class PixQrCodeParseError extends Error {
    constructor(message) {
        super(message);
        this.name = 'PixQrCodeParseError';
    }
}
exports.PixQrCodeParseError = PixQrCodeParseError;
function isTwoDigitField(value) {
    return /^\d{2}$/.test(value);
}
function isMerchantAccountInformationTag(tag) {
    if (!isTwoDigitField(tag))
        return false;
    const tagNumber = Number(tag);
    return tagNumber >= 26 && tagNumber <= 51;
}
function isTemplateTag(tag) {
    return isMerchantAccountInformationTag(tag) || tag === '62' || tag === '64';
}
function getTagName(tag, parent) {
    if (!parent) {
        if (isMerchantAccountInformationTag(tag)) {
            return 'Merchant Account Information';
        }
        return ROOT_NAMES[tag];
    }
    if (isMerchantAccountInformationTag(parent)) {
        return MAI_NAMES[tag];
    }
    if (parent === '62') {
        return ADD_DATA_NAMES[tag];
    }
    return undefined;
}
function assertRange(data, start, end) {
    if (!Number.isInteger(start) || !Number.isInteger(end)) {
        throw new PixQrCodeParseError('Os indices de parsing precisam ser inteiros.');
    }
    if (start < 0 || end < start || end > data.length) {
        throw new PixQrCodeParseError('Intervalo de parsing invalido para o BR Code informado.');
    }
}
function parseTLV(data, start = 0, end = data.length, parent = null) {
    if (typeof data !== 'string') {
        throw new PixQrCodeParseError('O BR Code precisa ser uma string.');
    }
    assertRange(data, start, end);
    const out = [];
    let i = start;
    while (i < end) {
        if (i + 4 > end) {
            throw new PixQrCodeParseError(`Campo TLV incompleto na posicao ${i}.`);
        }
        const tag = data.slice(i, i + 2);
        const rawLength = data.slice(i + 2, i + 4);
        if (!isTwoDigitField(tag)) {
            throw new PixQrCodeParseError(`Tag TLV invalida na posicao ${i}: "${tag}".`);
        }
        if (!isTwoDigitField(rawLength)) {
            throw new PixQrCodeParseError(`Tamanho TLV invalido na posicao ${i + 2}: "${rawLength}".`);
        }
        const length = Number(rawLength);
        const valueStart = i + 4;
        const valueEnd = valueStart + length;
        if (valueEnd > end) {
            throw new PixQrCodeParseError(`Valor TLV truncado para a tag ${tag}: esperado ${length} caracteres.`);
        }
        const value = data.slice(valueStart, valueEnd);
        const element = {
            tag,
            length,
            value,
            tagName: getTagName(tag, parent)
        };
        if (isTemplateTag(tag)) {
            element.children = parseTLV(value, 0, value.length, tag);
        }
        out.push(element);
        i = valueEnd;
    }
    return out;
}
function extractPixUrl(brcode) {
    const tlvs = parseTLV(brcode);
    const merchantAccounts = tlvs.filter((tlv) => isMerchantAccountInformationTag(tlv.tag) && tlv.children);
    const pixMerchantAccount = merchantAccounts.find((tlv) => tlv.children?.some((child) => child.tag === '00' && child.value.toUpperCase() === PIX_GUI));
    const merchantAccount = pixMerchantAccount ?? merchantAccounts[0];
    const urlField = merchantAccount?.children?.find((child) => child.tag === '25');
    return urlField?.value;
}
function decodeJwt(token) {
    const parts = token.trim().split('.');
    if (parts.length !== 3 || parts.some((part) => part.length === 0)) {
        throw new Error('Token JWT inesperado ou malformado.');
    }
    const header = parseJwtHeader(parts[0]);
    const payload = parseJwtPayload(parts[1]);
    return {
        header,
        payload,
        signature: parts[2]
    };
}
async function getDecodedPixJwt(brcode, options = {}) {
    const raw = extractPixUrl(brcode);
    if (!raw) {
        throw new Error('URL do payload nao encontrada no BR Code.');
    }
    const url = normalizePayloadUrl(raw);
    const fetchFn = options.fetch ?? globalThis.fetch;
    if (typeof fetchFn !== 'function') {
        throw new Error('Fetch API indisponivel. Use Node.js 18+ ou informe options.fetch.');
    }
    const controller = new AbortController();
    const timeoutMs = options.timeoutMs ?? 5_000;
    const timeout = setTimeout(() => controller.abort(), timeoutMs);
    try {
        const response = await fetchFn(url, {
            headers: {
                Accept: 'application/jwt, text/plain',
                ...options.headers
            },
            signal: controller.signal
        });
        if (!response.ok) {
            throw new Error(`Falha ao baixar payload Pix: HTTP ${response.status}.`);
        }
        const token = await response.text();
        return decodeJwt(token);
    }
    finally {
        clearTimeout(timeout);
    }
}
function normalizePayloadUrl(rawUrl) {
    const trimmed = rawUrl.trim();
    if (!trimmed) {
        throw new Error('URL do payload Pix vazia.');
    }
    const withProtocol = /^https?:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`;
    const url = new URL(withProtocol);
    if (url.protocol !== 'http:' && url.protocol !== 'https:') {
        throw new Error(`Protocolo nao suportado para payload Pix: ${url.protocol}`);
    }
    url.searchParams.set('DPP', getTodayIsoDate());
    return url.toString();
}
function getTodayIsoDate() {
    const now = new Date();
    const year = now.getFullYear();
    const month = String(now.getMonth() + 1).padStart(2, '0');
    const day = String(now.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
}
function parseJwtHeader(encoded) {
    const decoded = parseJwtJson(encoded);
    if (!isRecord(decoded)) {
        throw new Error('Header JWT inesperado ou malformado.');
    }
    return decoded;
}
function parseJwtPayload(encoded) {
    const text = base64UrlDecode(encoded);
    try {
        const decoded = JSON.parse(text);
        return isRecord(decoded) ? decoded : text;
    }
    catch {
        return text;
    }
}
function parseJwtJson(encoded) {
    try {
        return JSON.parse(base64UrlDecode(encoded));
    }
    catch {
        throw new Error('Token JWT inesperado ou malformado.');
    }
}
function base64UrlDecode(value) {
    const base64 = value.replace(/-/g, '+').replace(/_/g, '/');
    const padded = base64.padEnd(base64.length + ((4 - (base64.length % 4)) % 4), '=');
    const binary = globalThis.atob(padded);
    const bytes = Uint8Array.from(binary, (char) => char.charCodeAt(0));
    return new TextDecoder().decode(bytes);
}
function isRecord(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
